### PR TITLE
 remove DockerVolume from pod spec

### PR DESCRIPTION
**Pull request by gao-feng**
remove DockerVolume from pod spec

based on https://github.com/hyperhq/runv/pull/176

### CODE REVIEW

File: daemon/pod.go [hunk 1]
```diff
@@ -536,7 +537,7 @@ func (p *Pod) setupServices() error {
 	return err
 }
 
-// PrepareEtcHosts sets /etc/hosts for each container
+// Setu[EtcHosts sets /etc/hosts for each container
 func (p *Pod) setupEtcHosts() (err error) {
```
Comment (laijs) [inline daemon/pod.go]: setup

File: daemon/pod.go [hunk 2]
```diff
@@ -728,35 +729,66 @@ func (p *Pod) setupMountsAndFiles(sd Storage) (err error) {
...
-		p.volumes = append(p.volumes, vol)
+		if vol, ok := p.volumes[v.Name]; ok {
+			vol.Filepath = volInfo.Filepath
+			vol.Fstype = volInfo.Fstype
+			vol.Format = volInfo.Format
+			continue
+		}
+
+		p.volumes[v.Name] = volInfo
+	}
```
Comment (laijs) [inline daemon/pod.go]: I think only `p.volumes[v.Name] = volInfo` is Ok,  we don't need to copy it one field by field if the volume existed.

Comment (laijs) [inline daemon/pod.go]: It copies and keep DockerVolume. ignore my comments.

File: daemon/storage.go [hunk 3]
```diff
@@ -40,6 +40,7 @@ type Storage interface {
 
 	PrepareContainer(id, sharedir string) (*hypervisor.ContainerInfo, error)
 	InjectFile(src io.Reader, containerId, target, rootDir string, perm, uid, gid int) error
+	// CreateVolume should check if volume is existed to avoid allocating volume mulit time
 	CreateVolume(daemon *Daemon, podId, shortName string) (*hypervisor.VolumeInfo, error)
```
Comment (laijs) [inline daemon/storage.go]: does all the store drivers obeyed it?

File: daemon/pod.go [hunk 4]
```diff
@@ -422,7 +418,8 @@ func (p *Pod) createVolumes(daemon *Daemon) error {
 
 	sd := daemon.Storage
 	for i := range p.spec.Volumes {
-		if p.spec.Volumes[i].Source == "" {
+		// hyperd doesn't 'remove' vfs volume
+		if p.spec.Volumes[i].Driver != "vfs" {
 			vol, err = sd.CreateVolume(daemon, p.id, p.spec.Volumes[i].Name)
```
Comment (gnawux) [inline daemon/pod.go]: I think this patch changed the semantics of the `CreateVolume`. In the previous code, the daemon check the user spec, and only create volume when the user doesn't specify the source of the volume.

**Pull request merged**